sweep/fee_bumper.go

```diff
@@ -0,0 +1,623 @@
+package sweep
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/chainntnfs"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/lnutils"
+	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/lnwallet/chainfee"
+)
+
+const (
+	// NoHeightInfo is used when there's no height information available.
+	NoHeightInfo = math.MaxInt32
+
+	// defaultMaxFeeRatio is used when the max fee ratio is not set for the
+	// requested input.
+	defaultMaxFeeRatio = 0.1
+)
+
+var (
+	// ErrConfTargetTooSmall is used when the requested conf target is less
+	// than 1.
+	ErrConfTargetTooSmall = errors.New("conf target is too small")
+
+	// ErrInvalidStrategy specifies an invalid strategy.
+	ErrInvalidStrategy = errors.New("invalid strategy")
+
+	// ErrAlreadyMonitored is used when the requested tx has already been
+	// monitored.
+	ErrAlreadyMonitored = errors.New("record already monitored")
+
+	// ErrConflictArgs is used when conflicting arguments are both
+	// set/unset when making a request.
+	ErrConflictArgs = errors.New("conflict arguments used")
+)
+
+// Bumper helps managing the fee bumping of a given outpoint.
+type Bumper interface {
+	// Start starts the fee bumper service.
+	Start()
+
+	// Stop stops the fee bumper service.
+	Stop()
+
+	// RequestMonitor sends a request to monitor a given outpoint for
+	// potential fee bumping.
+	RequestMonitor(req *MonitorRequest) error
+
+	// NotifyConfirm specifies a previously requested outpoint has been
+	// swept.
+	NotifyConfirm(op *wire.OutPoint)
+}
+
+// Strategy specifies how the fee bump is performed.
+type Strategy uint8
```

Comment:
A slightly different way to represent this would be to have `Strategy` itself be an interface with methods for: seeing it things need to be bumped, and suggesting the proper fee rate. This would allow us to further encapsulate the logic of each strategy, and each strategy can also maintain its own state to make better decisions.